pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,316 @@
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileControllerRegistration(ctx context.Context, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	ctrlReg := &gardencorev1beta1.ControllerRegistration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: extension.Name,
+		},
+	}
+
+	regMutateFn := func() error {
+		// handle well known annotations
+		if v, ok := extension.Annotations[v1beta1constants.AnnotationPodSecurityEnforce]; ok {
+			metav1.SetMetaDataAnnotation(&ctrlReg.ObjectMeta, v1beta1constants.AnnotationPodSecurityEnforce, v)
+		} else {
+			delete(ctrlReg.Annotations, v1beta1constants.AnnotationPodSecurityEnforce)
+		}
+
+		ctrlReg.Spec = gardencorev1beta1.ControllerRegistrationSpec{
+			Resources: extension.Spec.Resources,
+			Deployment: &gardencorev1beta1.ControllerRegistrationDeployment{
+				Policy: extension.Spec.Deployment.ExtensionDeployment.Policy,
+				DeploymentRefs: []gardencorev1beta1.DeploymentRef{
+					{
+						Name: extension.Name,
+					},
+				},
+			},
+		}
+		return nil
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, gardenClient, ctrlReg, regMutateFn); err != nil {
+		return fmt.Errorf("failed to create or update ControllerRegistration: %w", err)
+	}
+	return nil
+}
+
+func (r *Reconciler) delete(ctx context.Context, log logr.Logger, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	conditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+
+	err := r.deleteVirtualClusterResources(ctx, log, gardenClient, extension)
+	if err != nil {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionDeleteFailed, err.Error())
+	} else {
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionDeleteSuccessful, "successfully deleted virtual cluster resources")
+	}
+
+	if err := errors.Join(err, r.updateExtensionStatus(ctx, log, extension, conditions)); err != nil {
+		return err
+	}
+
+	return r.removeFinalizers(ctx, log, extension)
+}
+
+func (r *Reconciler) deleteVirtualClusterResources(ctx context.Context, log logr.Logger, gardenClient client.Client, extension *operatorv1alpha1.Extension) error {
+	log.Info("Deleting extension virtual resources", "name", extension.Name)
+	var (
+		ctrlDeploy = &gardencorev1.ControllerDeployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: extension.Name,
+			}}
+
+		ctrlReg = &gardencorev1beta1.ControllerRegistration{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: extension.Name,
+			},
+		}
+	)
+
+	// deleting the controller deployment first to set the termination timestamp on the object. The deletion will be complete once the
+	// controllerRegistration has been deleted.
+	log.Info("Deleting controller deployment for extension", "extension", extension.Name)
```

Comment:
```suggestion
	log.Info("Deleting controller deployment for extension")
```

This should already be part of the logger. Remove also elsewhere